pkg/util/strategicpatch/patch.go

```diff
@@ -0,0 +1,454 @@
+	}
+
+	result, err := mergeMap(o, p, t)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(result)
+}
+
+const specialKey = "__patch"
+
+// Merge fields from a patch map into the original map.
+func mergeMap(original, patch map[string]interface{}, t reflect.Type) (map[string]interface{}, error) {
+	// If the map contains "__patch: strict", don't merge it, just use the
+	// patch map directly. Later on, can add a non-recursive strict that only
+	// affects the map that the __patch is in.
+	if v, ok := patch[specialKey]; ok {
+		if v == "strict" {
+			delete(patch, specialKey)
+			return patch, nil
+		}
+	}
+
+	// Start merging the patch into the original.
+	for k, patchV := range patch {
+		// If the value of this key is null, delete the key if it exists in the
+		// original. Otherwise, skip it.
+		if patchV == nil {
+			if _, ok := original[k]; ok {
+				delete(original, k)
+			}
+			continue
+		}
+
+		_, ok := original[k]
+		if !ok {
+			// If it's not in the original document, just take the patch value.
+			if original == nil {
+				original = map[string]interface{}{}
+			}
+			original[k] = patchV
+		} else {
+			// If they're both maps or lists, recurse into the value.
+			// First find the fieldPatchStrategy and fieldPatchMergeKey.
+			fieldType, fieldPatchStrategy, fieldPatchMergeKey, err := lookupPatchMetadata(t, k)
+			if err != nil {
+				return nil, err
+			}
+
+			originalType := reflect.TypeOf(original[k])
+			patchType := reflect.TypeOf(patchV)
+			if originalType == patchType {
+				if originalType.Kind() == reflect.Map && fieldPatchStrategy != "replace" {
+					typedOriginal := original[k].(map[string]interface{})
+					typedPatch := patchV.(map[string]interface{})
+					var err error
+					original[k], err = mergeMap(typedOriginal, typedPatch, fieldType)
+					if err != nil {
+						return nil, err
+					}
+					continue
+				} else if originalType.Kind() == reflect.Slice && fieldPatchStrategy == "merge" {
+					elemType := fieldType.Elem()
+					typedOriginal := original[k].([]interface{})
+					typedPatch := patchV.([]interface{})
+					var err error
+					original[k], err = mergeSlice(typedOriginal, typedPatch, elemType, fieldPatchMergeKey)
+					if err != nil {
+						return nil, err
+					}
+					continue
+				}
+			}
+
+			original[k] = patchV
+		}
+	}
+
+	return original, nil
+}
+
+func mergeSlice(original, patch []interface{}, elemType reflect.Type, mergeKey string) ([]interface{}, error) {
+	if len(original) == 0 && len(patch) == 0 {
+		return original, nil
+	}
+	// All the values must be of the same type, but not a list.
+	t, err := sliceElementType(original, patch)
+	if err != nil {
+		return nil, fmt.Errorf("types of list elements need to be the same, type: %s: %v",
+			elemType.Kind().String(), err)
+	}
+	if t.Kind() == reflect.Slice {
+		return nil, fmt.Errorf("not supporting merging lists of lists yet")
```

Comment:
In the struct tag model, you'd probably just put patchStrategy:merge on a list of list and it would just handle it correctly, and the mergeKey would just apply to the inner list vs the outer list.